modyn/trainer_server/internal/dataset/online_dataset.py

```diff
@@ -125,11 +126,25 @@ def _init_transforms(self) -> None:
         self._transform = self._bytes_parser_function
         self._setup_composed_transform()
 
-    def _init_grpc(self) -> None:
-        self._storage_channel = grpc.insecure_channel(self._storage_address, options=grpc_common_config())
-        if not grpc_connection_established(self._storage_channel):
-            raise ConnectionError(f"Could not establish gRPC connection to storage at address {self._storage_address}.")
-        self._storagestub = StorageStub(self._storage_channel)
+    def _init_grpc(self, worker_id: Optional[int]) -> None:
+        max_retries = 5
+        retry_delay = 1  # seconds
+
+        for attempt in range(1, max_retries + 1):
```

Comment:
It's okay this time, generally, this could be achieved through dedicated libraries like [tenacity](https://tenacity.readthedocs.io/en/latest/) that don't obfuscate the business logic that much :)

Reply:
Thanks for yet another nice pointer into the Python ecosystem :D Yeah we have such logic in many places, I guess we should refactor this consistently...